tensornetwork/contractors/stochastic_contractor.py

```diff
@@ -0,0 +1,144 @@
+# Copyright 2019 The TensorNetwork Authors
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#      http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+"""Stochastic Network Contraction."""
+
+from __future__ import absolute_import
+from __future__ import division
+from __future__ import print_function
+
+import random
+import numpy as np
+from tensornetwork import network, network_components
+from typing import Tuple, Set, Optional, Dict
+
+
+def find_parallel(edge: network_components.Edge) -> Tuple[Set, int]:
+  """Finds all edges shared between the nodes connected with the given edge.
+
+  Args:
+    edge: A non-dangling edge between two different nodes.
+
+  Returns:
+    parallel_edges: Edges that are parallel to the given edge.
+    parallel_dim: Product of sizes of all parallel edges.
+  """
+  if edge.is_dangling():
+    raise ValueError(
+             "Cannot find parallel edges for dangling edge {}".format(edge))
+  nodes = {edge.node1, edge.node2}
+  parallel_dim = 1
+  parallel_edges = set()
+  for edge in edge.node1.edges:
+    if set(edge.get_nodes()) == nodes:
+      parallel_edges.add(edge)
+      edge_size = edge.node1.get_tensor().shape.as_list()[edge.axis1]
```

Comment:
This reminds me, we should add a `node.shape` property to support multi backends.

No changes here for now though.